WalletWasabi/Services/Terminate/TerminateService.cs

```diff
@@ -0,0 +1,90 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Logging;
+
+namespace WalletWasabi.Services.Terminate
+{
+	public class TerminateService
+	{
+		private Func<Task> _terminateApplicationAsync;
+
+		private const long TerminateStatusIdle = 0;
+		private const long TerminateStatusInProgress = 1;
+		private const long TerminateFinished = 2;
+
+		private long _terminateStatus;
+
+		public TerminateService(Func<Task> terminateApplicationAsync)
+		{
+			_terminateApplicationAsync = terminateApplicationAsync;
+			AppDomain.CurrentDomain.ProcessExit += CurrentDomain_ProcessExit;
+			Console.CancelKeyPress += Console_CancelKeyPress;
+		}
+
+		private void CurrentDomain_ProcessExit(object? sender, EventArgs e)
+		{
+			Logger.LogDebug("ProcessExit was called.");
+
+			// This must be a blocking call because after this the OS will terminate Wasabi process if exists.
+			Terminate();
+		}
+
+		private void Console_CancelKeyPress(object? sender, ConsoleCancelEventArgs e)
+		{
+			Logger.LogWarning("Process was signaled for termination.");
+			// This must be a blocking call because after this the OS will terminate Wasabi process if exists.
+			// In some cases CurrentDomain_ProcessExit is called after this by the OS.
+			Terminate();
+		}
+
+		/// <summary>
+		/// This will terminate the application. This is a blocking method and no return after this call as it will exit the application.
+		/// </summary>
+		public void Terminate(int exitCode = 0)
+		{
+			var prevValue = Interlocked.CompareExchange(ref _terminateStatus, TerminateStatusInProgress, TerminateStatusIdle);
+			Logger.LogTrace($"Terminate was called from ThreadId: {Thread.CurrentThread.ManagedThreadId}");
+			if (prevValue != TerminateStatusIdle)
+			{
+				// Secondary callers will be blocked until the end of the termination.
+				while (_terminateStatus != TerminateFinished)
+				{
+				}
```

Comment:
This busy waiting does not make any sense to me. You force CPU to work with no clear benefit.

Typical scenario will be: 

* First thread calls this method.
* Second thread calls this method and you start busy waiting.
* First thread gets to Environment.Exit() call which terminates the application immediately (no matter what the second thread does).

I would say it makes more sense to use Thread.Sleep() than using busy wait here. What about simply logging that the application is already stopping and returning back? Is there any problem with that?

Reply:
I added sleep to let CPU rest for a while. 

Why blocking?

- CurrentDomain_ProcessExit called on OS thread and it should not return until the end of the termination. If it would the OS would have killed the application - according to my tests.
- The Main function should not return until proper termination. 
- Console_CancelKeyPress the same applies as before, however, it could be solved e.Cancel = true and return but it makes no sense for me. That was used before and the only benefit was that the user can press ctrl-c more times and it was indicated in the log.